src/Simplex/FileTransfer/Protocol.hs

```diff
@@ -1 +1,364 @@
+
+data FileCmdTag = forall p. FilePartyI p => CT (SFileParty p) (FileCommandTag p)
+
+instance FilePartyI p => Encoding (FileCommandTag p) where
+  smpEncode = \case
+    NEW_ -> "NEW"
+    ADD_ -> "ADD"
+    UP_ -> "UP"
+    SDEL_ -> "SDEL"
+    DOWN_ -> "DOWN"
+    RDEL_ -> "RDEL"
+    PING_ -> "PING"
+  smpP = messageTagP
+
+instance Encoding FileCmdTag where
+  smpEncode (CT _ t) = smpEncode t
+  smpP = messageTagP
+
+instance ProtocolMsgTag FileCmdTag where
+  decodeTag = \case
+    "NEW" -> Just $ CT SSender NEW_
+    "ADD" -> Just $ CT SSender ADD_
+    "UP" -> Just $ CT SSender UP_
+    "SDEL" -> Just $ CT SSender SDEL_
+    "DOWN" -> Just $ CT SRecipient DOWN_
+    "RDEL" -> Just $ CT SRecipient RDEL_
+    "PING" -> Just $ CT SRecipient PING_
+    _ -> Nothing
+
+
+instance FilePartyI p => ProtocolMsgTag (FileCommandTag p) where
+  decodeTag s = decodeTag s >>= (\(CT _ t) -> checkParty' t)
+
+{- newtype NtfRegCode = NtfRegCode ByteString
+  deriving (Eq, Show)
+
+instance Encoding NtfRegCode where
+  smpEncode (NtfRegCode code) = smpEncode code
+  smpP = NtfRegCode <$> smpP
+
+instance StrEncoding NtfRegCode where
+  strEncode (NtfRegCode m) = strEncode m
+  strDecode s = NtfRegCode <$> strDecode s
+  strP = NtfRegCode <$> strP
+  
+instance FromJSON NtfRegCode where
+  parseJSON = strParseJSON "NtfRegCode"
+
+instance ToJSON NtfRegCode where
+  toJSON = strToJSON
+  toEncoding = strToJEncoding
+
+data NewNtfEntity (e :: NtfEntity) where
+  NewNtfTkn :: DeviceToken -> C.APublicVerifyKey -> C.PublicKeyX25519 -> NewNtfEntity 'Token
+  NewNtfSub :: NtfTokenId -> SMPQueueNtf -> NtfPrivateSignKey -> NewNtfEntity 'Subscription
+
+deriving instance Show (NewNtfEntity e)
+
+data ANewNtfEntity = forall e. NtfEntityI e => ANE (SNtfEntity e) (NewNtfEntity e)
+
+deriving instance Show ANewNtfEntity
+
+instance NtfEntityI e => Encoding (NewNtfEntity e) where
+  smpEncode = \case
+    NewNtfTkn tkn verifyKey dhPubKey -> smpEncode ('T', tkn, verifyKey, dhPubKey)
+    NewNtfSub tknId smpQueue notifierKey -> smpEncode ('S', tknId, smpQueue, notifierKey)
+  smpP = (\(ANE _ c) -> checkEntity c) <$?> smpP
+
+instance Encoding ANewNtfEntity where
+  smpEncode (ANE _ e) = smpEncode e
+  smpP =
+    A.anyChar >>= \case
+      'T' -> ANE SToken <$> (NewNtfTkn <$> smpP <*> smpP <*> smpP)
+      'S' -> ANE SSubscription <$> (NewNtfSub <$> smpP <*> smpP <*> smpP)
+      _ -> fail "bad ANewNtfEntity"
+
+   -}
+
+instance Protocol FileResponse where
+  type ProtoCommand FileResponse = FileCmd
+  type ProtoType FileResponse = 'PNTF
+  protocolClientHandshake = ntfClientHandshake
+  protocolPing = FileCmd SRecipient PING
+  protocolError = \case
+    FRErr e -> Just e
+    _ -> Nothing
+
+data FileCommand (p :: FileParty) where
+  -- Sender key, recipients keys, chunk size
+  NEW :: SndPublicVerifyKey -> [RcvPublicVerifyKey] -> Int -> FileCommand Sender
+  ADD :: [RcvPublicVerifyKey] -> FileCommand Sender
+  UP :: FileCommand Sender
```

Comment:
the idea here is that attachment would be somewhere out of band, like HTTP2 body part... we will see, ok for now.